File: solutions/0/0/4/7/p0047_s0.c
/* Solution to Euler Problem 47: Distinct Primes Factors. */
#include "runner.h"

/* Count distinct prime factors of n by trial division with in-place reduction; O(sqrt(n)). */
static int prime_factor_count(int n) {
    if (n <= 1) return 0;
    int count = 0;
    int d = 2;
    while (d * d <= n) {
        if (n % d == 0) {
            count++;
            /* Strip every copy of d so later composite d cannot divide; each divisor found is prime. */
            while (n % d == 0) {
                n /= d;
            }
        }
        d++;
    }
    /* Any remainder above 1 is a prime factor exceeding the square root of the original n. */
    if (n > 1) count++;
    return count;
}

/* Linear scan for the first run of n consecutive integers each with n distinct prime factors.
   Each count is O(sqrt(k)) trial division; the window test short-circuits on the first failure,
   giving roughly O(answer * sqrt(answer)) overall. */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int n = parse_int(argv[1]);
    int number = 2;
    while (1) {
        int all_match = 1;
        for (int i = 0; i < n; i++) {
            if (prime_factor_count(number + i) != n) {
                all_match = 0;
                break;
            }
        }
        if (all_match) { snprintf(_answer, sizeof _answer, "%lld", (long long)((long long)number)); return _answer; }
        number++;
    }
}

File: solutions/public/p0047/p0047_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 47: Distinct Primes Factors [Level 1]. """
from __future__ import annotations

import itertools

from solver.runners import runner


def prime_factorization(n: int) -> tuple[tuple[int, int], ...]:
    """Factor n by trial division with in-place reduction, returning (prime, exponent) pairs; O(sqrt(n))."""
    if n <= 1:
        return ()
    factors = []
    d = 2
    while d * d <= n:
        if n % d == 0:
            exp = 0
            while n % d == 0:
                n //= d
                exp += 1
            factors.append((d, exp))
        d += 1
    if n > 1:
        factors.append((n, 1))
    return tuple(factors)


def prime_factor_count(n: int) -> int:
    """Number of distinct prime factors of n (the count of factorization pairs)."""
    return len(prime_factorization(n))


@runner.main
def solve(*args: str) -> str:
    """Linear scan for the first run of n consecutive integers each with n distinct prime factors.

    Each count is O(sqrt(k)) by trial division; the window test short-circuits via any() on the
    first failure, so the search costs about O(answer * sqrt(answer)).
    """
    n = runner.parse_int(args[0])

    return str(next(
        (number for number in itertools.count(2) if not any((prime_factor_count(number + i) != n for i in range(0, n))))
    ))


if __name__ == "__main__":
    raise SystemExit(solve())
